src/hotspot/share/gc/shared/memAllocator.cpp

```diff
@@ -376,12 +372,10 @@ oop MemAllocator::allocate() const {
   return obj;
 }
 
-void MemAllocator::mem_clear(HeapWord* mem) const {
+void MemAllocator::mem_clear(HeapWord* mem, size_t hdr_size_bytes) const {
   assert(mem != nullptr, "cannot initialize null object");
-  const size_t hs = oopDesc::header_size();
-  assert(_word_size >= hs, "unexpected object size");
-  oopDesc::set_klass_gap(mem, 0);
-  Copy::fill_to_aligned_words(mem + hs, _word_size - hs);
+  assert(_word_size * HeapWordSize >= hdr_size_bytes, "unexpected object size");
+  Copy::fill_to_bytes((char*)mem + hdr_size_bytes, _word_size * HeapWordSize - hdr_size_bytes);
```

Comment:
Performance-sensitive path, let's keep `Copy::fill_to_aligned_words` and compensate for misalignment with a single store, as we do everywhere else.